ui/src/criteria/concept.tsx

```diff
@@ -171,48 +201,67 @@ function ConceptEdit(props: ConceptEditProps) {
 
             // Copy over existing children in case they're being loaded in
             // parallel.
+            let childChildren = data[id]?.children;
+            if (!childChildren) {
+              if (!hierarchy) {
+                hasChildren = !!entity.listChildren && !listChildrenEntity;
+              }
+              childChildren = !hasChildren ? [] : undefined;
+            }
+
             data[id] = {
               data: row,
-              children: data[id]?.children || (!hasChildren ? [] : undefined),
+              children: childChildren,
             };
           });
         }
 
-        if (id) {
+        if (parentId) {
           // Store children even if the data isn't loaded yet.
-          data[id] = { ...data[id], children };
+          data[parentId] = { ...data[parentId], children };
         } else {
-          data.root = { children, data: {} };
+          data.root = {
+            children: [...(data?.root?.children || []), ...children],
+            data: {
+              [ENTITY_ATTRIBUTE]: entity.name,
+            },
+          };
         }
       });
     },
     []
   );
 
-  const conceptsState = useAsyncWithApi<void>(
-    useCallback(() => {
-      updateData(() => ({}));
-      return api
-        .searchEntityInstances(
-          searchFilter(
-            props.data,
-            props.cohort.underlayName,
-            !hierarchyPath ? query : ""
+  const fetch = useCallback(() => {
+    updateData(() => ({}));
+    return Promise.all(
+      props.data.entities
+        .filter((entity) => !hierarchy || entity.name === hierarchy.entity)
+        .map((entity) =>
+          api.searchEntityInstances(
+            searchRequest(
+              props.data.columns,
+              entity,
+              props.cohort.underlayName,
+              !hierarchy ? query : ""
+            )
           )
         )
-        .then((res) => {
-          processEntities(res);
-        });
-    }, [
-      api,
-      props.data.entity,
-      props.cohort.underlayName,
-      processEntities,
-      hierarchyPath,
-      query,
-    ]),
-    hierarchyPath
-  );
+    ).then((res) => {
+      res?.forEach((r, i) =>
+        processEntities(r, props.data.entities[i], hierarchy)
+      );
+    });
+  }, [
+    api,
+    props.data.columns,
+    props.data.entities,
+    props.cohort.underlayName,
+    processEntities,
+    hierarchy,
+    query,
+  ]);
+  const conceptsState = useAsyncWithApi<void>(fetch, fetch);
```

Comment:
Is this second parameter correct?

Reply:
Yes, but I'm going to remove it. I realized it was causing two layers of equality checking so I'm going to change it to always use the callback as the watch value instead of passing one in separately. I've added a TODO for now.